python-dice-roll/source_code_step_3/dice.py

```diff
@@ -0,0 +1,170 @@
+    ),
+    4: (
+        "┌─────────┐",
+        "│  ●   ●  │",
+        "│         │",
+        "│  ●   ●  │",
+        "└─────────┘",
+    ),
+    5: (
+        "┌─────────┐",
+        "│  ●   ●  │",
+        "│    ●    │",
+        "│  ●   ●  │",
+        "└─────────┘",
+    ),
+    6: (
+        "┌─────────┐",
+        "│  ●   ●  │",
+        "│  ●   ●  │",
+        "│  ●   ●  │",
+        "└─────────┘",
+    ),
+}
+DIE_HEIGHT = len(DICE_ART[1])
+DIE_WIDTH = len(DICE_ART[1][0])
+DIE_FACE_SEPARATOR = " "
+
+
+def parse_input(input_string):
+    """Return `input_string` as an integer between 1 and 6.
+
+    Check if `input_string` is an integer number between 1 and 6.
+    If so, return an integer with the same value. Otherwise, tell
+    the user to enter a valid number and quit the program.
+    """
+    is_valid = input_string.isdigit() and 1 <= int(input_string) <= 6
+    if is_valid:
+        return int(input_string)
+    else:
+        print("Please enter an integer between 1 and 6.")
+        raise SystemExit(1)
+
+
+def roll_dice(num_dice):
+    """Return a list of integers with length `num_dice`.
+
+    Each integer in the returned list is a random number between
+    1 and 6, inclusive.
+    """
+    roll_results = []
+    for _ in range(num_dice):
+        roll = random.randint(1, 6)
+        roll_results.append(roll)
+    return roll_results
+
+
+def generate_dice_faces_diagram(dice_values):
+    """Return ASCII die faces diagram for `dice_values`.
+
+    The string returned contains an ASCII representation of each die.
+    For example, if `dice_values = [4, 1, 3, 2]` then the string
+    returned looks like this:
+
+    ~~~~~~~~~~~~~~~~~~~ RESULTS ~~~~~~~~~~~~~~~~~~~
+    ┌─────────┐ ┌─────────┐ ┌─────────┐ ┌─────────┐
+    │  ●   ●  │ │         │ │  ●      │ │  ●      │
+    │         │ │    ●    │ │    ●    │ │         │
+    │  ●   ●  │ │         │ │      ●  │ │      ●  │
+    └─────────┘ └─────────┘ └─────────┘ └─────────┘
+    """
+    # Generate a list of dice faces from DICE_ART
+    dice_faces = []
+    for value in dice_values:
+        dice_faces.append(DICE_ART[value])
+
+    # Generate a list containing the dice faces rows
+    dice_faces_rows = []
+    for row_idx in range(DIE_HEIGHT):
+        row_components = []
+        for die in dice_faces:
+            row_components.append(die[row_idx])
+        row_string = DIE_FACE_SEPARATOR.join(row_components)
+        dice_faces_rows.append(row_string)
+
+    # Generate header with the word "RESULTS" centered
+    width = len(dice_faces_rows[0])
+    diagram_header = " RESULTS ".center(width, "~")
+
+    dice_faces_diagram = "\n".join([diagram_header] + dice_faces_rows)
+    return dice_faces_diagram
+
+
+# EXERCISE:
+# Refactor generate_dice_faces_diagram() to satisfy the single-responsibility
+# principle. Take advantage of _get_dice_faces() and write another helper
+# function called _generate_dice_faces_rows() to extract the functionality
+# from lines 115 to 121.
```

Comment:
In case you don't agree with the step 4 refactoring mentioned elsewhere, then you'll have to make sure that the line numbers here make sense. Currently these commented-out instructions about refactoring point to lines 115 to 121, which isn't correct.